kernel/source/messaging/ipc.c

```diff
@@ -216,23 +248,390 @@ void ipc_irq(void)
+		ipc_circ_buf.waiting_in_buffer_cnt--;
+		ipc_circ_buf.had_relaunch_flag = 0;
+		write32(HW_IPC_ARMCTRL, (read32(HW_IPC_ARMCTRL) & 0x30) | ((ipc_circ_buf.waiting_in_buffer_cnt == (IPC_CIRCULAR_BUFFER_SIZE - 1) ? IPC_ARM_ACK_OUT : 0) | IPC_ARM_INCOMING));
+	}
+}
+
+static void DoFlushIpcRequestAfterInvalidate(IpcRequest *request)
+{
+	const u32 req_cmd = request->RequestCommand;
+	if (req_cmd == IOS_IOCTL)
+	{
+		DCFlushRange(request->Data.Ioctl.InputBuffer, request->Data.Ioctl.InputLength);
+		DCFlushRange(request->Data.Ioctl.IoBuffer, request->Data.Ioctl.IoLength);
+	}
+	else if (req_cmd == IOS_READ)
+	{
+		DCFlushRange(request->Data.Read.Data, request->Result);
+	}
+	else if (req_cmd == IOS_IOCTLV)
+	{
+		const u32 total_arg_count = request->Data.Ioctlv.InputArgc + request->Data.Ioctlv.IoArgc;
+		for(u32 i = 0; i < total_arg_count; ++i)
+		{
+			DCFlushRange(request->Data.Ioctlv.Data[i].Data, request->Data.Ioctlv.Data[i].Length);
+		}
+		DCFlushRange(request->Data.Ioctlv.Data, total_arg_count * sizeof(IoctlvMessageData));
+	}
+	
+	ipc_circ_buf.circular_backing_array[ipc_circ_buf.prepare_to_send_idx] = request;
+	ipc_circ_buf.prepare_to_send_idx = (ipc_circ_buf.prepare_to_send_idx + 1) % IPC_CIRCULAR_BUFFER_SIZE;
+	ipc_circ_buf.ready_to_send_cnt++;
+
+	DoSendIpcRequest();
+	return;
+}
+
+static int CheckAddrIsFine_Inner(const u32 addr, const u32 size)
+{
+	const u32 addr_end = addr + size;
+	return (addr < addr_end) && ((0x10000000 <= addr && addr_end < 0x13620000) || (addr_end < 0x01800000));
+}
+static int CheckAddrIsFine(const u32 addr, const u32 size)
+{
+	const int ret = CheckAddrIsFine_Inner(addr, size);
+	if(!ret)
+		printk("IPC: failed buf check: ptr=%08x len=%d\n", addr, size);
+	return ret;
+}
+static int CheckPtrIsFine(const void* const addr, const u32 size)
+{
+	return CheckAddrIsFine((u32)addr, size);
+}
+
+void IpcHandler(void)
+{
+	SetThreadPriority(0, 0x40);
+	IpcHandlerRequest.Command = IOS_INTERRUPT;
+
+	const s32 messageQueue = CreateMessageQueue((void**)IpcHandlerMessageQueueData, 50);
+	if (messageQueue < 0)
+		return;
+	
+	s32 ret = RegisterEventHandler(IRQ_IPC, messageQueue, &IpcHandlerRequest);
+	if(ret != IPC_SUCCESS)
+		return;
+
+	// enable incoming & outgoing ipc messages from PPC
+	if ((read32(HW_IPC_PPCCTRL) & 0x30) == 0)
+	{
+		write32(HW_IPC_PPCCTRL, 0);
+	}
+
+	ClearAndEnableIPCInterrupt();
+
+	while(1)
+	{
+		IpcMessage *message_ptr = NULL;
+
+		do {
+			ret = ReceiveMessage(messageQueue, (void**)&message_ptr, None);
+		} while (ret != IPC_SUCCESS);
+
+		while (1)
+		{
+			if(message_ptr->Request.Command == IOS_REPLY)
+			{
+				DoFlushIpcRequestAfterInvalidate(&message_ptr->Request);
+				break;
+			}
+
+			IpcMessage *ppc_message = (void*)read32(HW_IPC_PPCMSG);
+			write32(HW_IPC_PPCMSG, (u32)ppc_message);
+
+			if (message_ptr->Request.Command != IOS_INTERRUPT)
+			{
+				printk("UNKNOWN MESSAGE: %u\n", message_ptr->Request.Command);
+				ret = ReceiveMessage(messageQueue, (void**)&message_ptr, None);
```

Comment:
you might want to relook at the while loops in ghidra & do a refactor. in the other threads i was able to reduce them to 1 while loop

it also looks like IOS does a continue here, instead of repeating the receive message
i suspect a lot of compiler restructuring has happened from the original code